crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,264 @@
+use crate::schedule::{SystemDescriptor, SystemLabel, SystemSet};
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc, Mutex,
+    },
+};
+
+static NEXT_NODE_ID: AtomicU32 = AtomicU32::new(0);
+
+/// A builder for creating graphs of dependent parallel execution within a `SystemStage`.
+///
+/// # Sequential Execution
+/// The simplest graph is a sequence of systems that need to run one after another.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// let graph = SystemGraph::new();
+/// graph
+///   .root(sys_a.system())
+///   .then(sys_b.system())
+///   .then(sys_c.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+///
+/// # Fan Out
+/// `SystemGraphNode::then` can be called repeatedly on the same node to create multiple fan-out
+/// branches. All fanned out systems will not execute until the original has finished.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// let graph = SystemGraph::new();
+///
+/// let start_a = graph.root(sys_a.system());
+///
+/// start_a.then(sys_b.system());
+/// start_a.then(sys_c.system());
```

Comment:
I think we could probably improve on this API.
Instead of this, I'd want to do something like:

```rust
graph.root(sys_a.system()).then((sys_b.system(), sys_c.system()));
```
Then should be able to take 1 or more systems instead of having to store the variable and then call `then` twice (with-out chaining).

Also I could easily see a user getting confused on the difference between chaining `then` calls and then making a variable and calling `then` on the same thing twice.

Reply:
I'm actually a bit more confused by that example. Does calling with a tuple of systems return the node representing the last system in the chain? Does it return a tuple and fan-out? Can I input a iterator and get an iterator of fan-out systems?

I think it's worth splitting that out into it's own function like `join_into` rather than having a single entry point.

Reply:
Added a `split_into` implementation that does just this.